### DISCUSSION TITLE
non_argument_deps for Graph-Backed Assets

### DISCUSSION THREAD

**Discussion by khongtrunght** [dagster-io/dagster#12438] [Q&A] [answered]
I've only been using Dagster recently. I've read the documentation but haven't been able to figure out how to create non_argument_deps for Graph-Backed Assets. Is there anyone here with experience that can help me, thanks!

**Comment by sryza** [comment, marked as answer]
This is currently the way to do it:

```python
@asset
def upstream_asset() -> None:
    ...

@op(ins["start": In(dagster_type=Nothing)})
def my_op() -> None:
    ...

@graph
def my_graph(input1):
    return my_op(start=input1)

my_asset = AssetsDefinition.from_graph(my_graph, keys_by_input_name={"input1": "upstream_asset"})
```

IMO there's definitely opportunity to make this more ergonomic and discoverable.

**Reply by khongtrunght** [reply]
Thank you for your reply. But I have some issues with this approach. The error says that  graph **my_graph** has unmapped input **input1** . Remove it or pass it to the appropriate op/graph invocation.

**Reply by sryza** [reply]
Oops - I had a typo. Just updated the example.

**Reply by simonvanderveldt** [reply]
@sryza It seems like `@op` doesn't have a `dagster_type` argument? https://docs.dagster.io/_apidocs/ops#dagster.op

```
UserWarning: Error loading repository location non_argument.py:TypeError: op() got an unexpected keyword argument 'dagster_type
```

**Reply by sryza** [reply]
@simonvanderveldt my bad - another typo - updated the example again

**Reply by simonvanderveldt** [reply]
@sryza Still a couple of typos it looks like ;)
This worked for me

```
from dagster import asset, op, Nothing, graph, AssetsDefinition, In, AssetKey, Definitions

@asset
def upstream_asset() -> None:
    pass

@op(ins={"start": In(dagster_type=Nothing)})
def my_op() -> None:
    pass

@graph
def my_graph(input1):
    return my_op(start=input1)

my_asset = AssetsDefinition.from_graph(my_graph, keys_by_input_name={"input1": AssetKey("upstream_asset")})

defs = Definitions(assets=[my_asset])
```
